x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/upgrade_prebuilt_rules_table/use_filter_prebuilt_rules_to_upgrade.ts

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useMemo } from 'react';
+import type { RuleUpgradeInfoForReview } from '../../../../../../common/detection_engine/prebuilt_rules/api/review_rule_upgrade/response_schema';
+import type { FilterOptions } from '../../../../rule_management/logic/types';
+
+export type UpgradePrebuiltRulesTableFilterOptions = Partial<
+  Pick<FilterOptions, 'filter' | 'tags'>
+>;
+
+export const useFilterPrebuiltRulesToUpgrade = ({
```

Comment:
So I actually wanted to create a reusable hook for both adding and upgrading rules cause the filtering logic is the same. The problem is that the type of `rules` is slightly different in these two cases, because in the `install` case, all the properties of the rule (that I need to filter on) are top-level properties (i.e `name` and `tags`).

In the upgrade case, those are nested within `rule`:
```
export interface RuleUpgradeInfoForReview {
  id: RuleObjectId;
  rule_id: RuleSignatureId;
  rule: DiffableRule;
  diff: PartialRuleDiff;
  revision: number;
}
```
So I tried making it generic for both like:
```
export const useFilterPrebuiltRulesToInstallOrUpgrade = ({
  rules,
  filterOptions,
}: {
  rules: RuleInstallationInfoForReview[] | Array<RuleUpgradeInfoForReview['rule']>;
  filterOptions: AddOrUpgradePrebuiltRulesTableFilterOptions;
}) => {
  ...
```
but that means that the Rule Upgrade table will be populated by `DiffableRule`-type objects instead of `RuleUpgradeInfoForReview`, and that means additional changes among the context, which I think we don't want just to make this reusable.

I'd like to discuss this with you anyways to understand what are you thinking, but I'd rather merge this PR as is with two separate hooks and refactor next week after FF.